### PR TITLE
Telemetry rewrite

### PR DESCRIPTION
**Pull request by etnguyen03**
# Rewrite telemetry
Issue #7 

## Description
Rewrite telemetry to meet the specification as stated in the above issue, and implement Jason's error and logging classes. Meet the command specification as stated in command_ingest's README.

## Steps to Reproduce
1. Calls made to `telemetry.enqueue()` should enqueue the message (error, log, command) to `general_queue`.
2. The thread `decide()` should remove messages to `general_queue` as they come in and, depending on the format (error, log, command) send them to the error stack, log stack, or `command_ingest` respectively.
3. When `dump()` is called, subpackets from the error stack and the log stack (in that order) should be popped off their stacks, converted to their string representations, and concatenated until a maximum size is set (defined in config). This is then encoded to base64 and sent through the APRS (radio_output).
4. When `clear_buffers` is called, the two stacks and the general queue should be cleared.

## Affected Areas
* `core/error.py` (comments, stylistic changes)
* `core/log.py` (same)
* `submodules/radio_output/__init__.py` (fix the first import line)
* `submodules/telemetry/README.md` (readme)
* `submodules/telemetry/__init__.py`

### CODE REVIEW

File: submodules/telemetry/__init__.py [hunk 1]
```diff
@@ -1,166 +1,129 @@
...
 
 
 def start():
     """
     Starts the telemetry send thread
     """
-    t2 = ThreadHandler(target=partial(telemetry_send),
-                       name="telemetry-telemetry_send")
-    t2.start()
+    threadDecide = ThreadHandler(target=partial(decide()),
```
Comment (JayChen35) [inline submodules/telemetry/__init__.py]: The `partial()` function takes a function object, not a function call. 

Recommended Change:
`threadDecide = ThreadHandler(target=partial(decide), name="telemetry-decide")`

File: submodules/telemetry/__init__.py [hunk 2]
```diff
@@ -1,166 +1,129 @@
...
-    global event_packet_buffer, packet_lock
-    with packet_lock:
-        packet = "Z"
-        packet += str(base64.b64encode(struct.pack('d', time.time())))
-        packet += str(base64.b64encode(event))
-        event_packet_buffer.append(packet)
-
-
-def enqueue_submodule_packet(packet):
+@command("telem_dump")
```
Comment (JayChen35) [inline submodules/telemetry/__init__.py]: we are removing the `command` decorator with manually recording each function in a db in `command_ingest`

File: submodules/telemetry/__init__.py [hunk 3]
```diff
@@ -1,166 +1,129 @@
...
-    Immediately send telemetry packets in both telemetry and event packet queues
-    """
-    global telem_packet_buffer, event_packet_buffer
-    beg_count = len(telem_packet_buffer) + len(event_packet_buffer)
-    send()
-    logger.debug("Sent " + str(beg_count - len(telem_packet_buffer) -
-                               len(event_packet_buffer)) + " telemetry packets")
+logger = logging.getLogger("TELEMETRY")
 
+general_queue = collections.deque()
```
Comment (JayChen35) [inline submodules/telemetry/__init__.py]: By initializing these variables outside a function they will be reinitialized everytime the module is imported. Put these variables within the `start()` function and initialize them there. 
i.e.
```
def start():
  global general_queue
  general_queue = collections.deque()
```

File: submodules/telemetry/__init__.py [hunk 4]
```diff
@@ -1,166 +1,129 @@
...
+log_stack = collections.deque()
```
Comment (JayChen35) [inline submodules/telemetry/__init__.py]: Reference global variable initialization above

File: submodules/telemetry/__init__.py [hunk 5]
```diff
@@ -1,166 +1,129 @@
...
+err_stack = collections.deque()
```
Comment (JayChen35) [inline submodules/telemetry/__init__.py]: Reference global variable initialization above

File: submodules/telemetry/__init__.py [hunk 6]
```diff
@@ -1,166 +1,129 @@
...
 
-@command("telem_burst")
-def telemetry_burst_command():
-    """
-    Burst command; ignores isTJseen and other checks.
-    """
-    send(ignoreADCS=True)
-
+packet_lock = Lock()
```
Comment (JayChen35) [inline submodules/telemetry/__init__.py]: Reference global variable initialization above

File: submodules/telemetry/__init__.py [hunk 7]
```diff
@@ -1,166 +1,129 @@
...
+                command_ingest.enqueue(message[4:])
+            elif type(message) is error.Error:
+                err_stack.append(message)
+            elif type(message) is log.Log:
+                log_stack.append(message)
+            else:   # Shouldn't execute (enqueue() should catch it) but here just in case
+                logger.error("Message prefix invalid.")
 
 
 def start():
```
Comment (JayChen35) [inline submodules/telemetry/__init__.py]: Make sure you initialize ALL global variables here. (Refer above)

**Review by JayChen35 — COMMENTED** [review]
**LOVING** the documentation. Good Work. Made a few changes, resubmit and request when done!

**Review by abagali1 — CHANGES_REQUESTED** [review]
what jason said^

File: submodules/telemetry/__init__.py (hunk 1 above)
Comment (etnguyen03) [reply submodules/telemetry/__init__.py]: Done in 963480f

File: submodules/telemetry/__init__.py (hunk 2 above)
Comment (etnguyen03) [reply submodules/telemetry/__init__.py]: Removed all of them in 963480f (at least those in telemetry)

File: submodules/telemetry/__init__.py (hunk 3 above)
Comment (etnguyen03) [reply submodules/telemetry/__init__.py]: Done in 963480f

File: submodules/telemetry/__init__.py (hunk 4 above)
Comment (etnguyen03) [reply submodules/telemetry/__init__.py]: Done in 963480f

File: submodules/telemetry/__init__.py (hunk 5 above)
Comment (etnguyen03) [reply submodules/telemetry/__init__.py]: Done in 963480f

File: submodules/telemetry/__init__.py (hunk 6 above)
Comment (etnguyen03) [reply submodules/telemetry/__init__.py]: Done in 963480f

File: submodules/telemetry/__init__.py (hunk 7 above)
Comment (etnguyen03) [reply submodules/telemetry/__init__.py]: Done in 963480f

**Comment by etnguyen03** [comment]
Redid parts of telemetry, fixed several bugs I identified:

* Lengths of packets would sometimes be longer than the config's length (I'm checking the length of the squished packets, not the length of the B64'd packets)
* Not checking if there's stuff to deque before I pop

**Review by abagali1 — DISMISSED** [review]

**Review by abagali1 — APPROVED** [review]

**Pull request merged**